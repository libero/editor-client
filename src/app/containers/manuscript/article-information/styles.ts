import { makeStyles } from '@material-ui/core/styles';

export const useArticleInformationStyles = makeStyles((theme) => ({
  sectionContainer: {
    display: 'flex',
    alignItems: 'flex-start'
  },
  infoSection: {
    flex: 1,
    lineHeight: '24px',
    paddingTop: theme.spacing(0.5)
  },
  editButton: {
    flexGrow: 0,
    padding: 6
  }
}));
