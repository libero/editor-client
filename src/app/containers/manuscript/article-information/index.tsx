import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';
import { useSelector, useDispatch } from 'react-redux';

import { SectionContainer } from 'app/components/section-container';
import { getArticleInformation } from 'app/selectors/manuscript.selectors';
import { useArticleInformationStyles } from 'app/containers/manuscript/article-information/styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ArticleInfoFormDialog } from 'app/containers/article-info-form-dialog';

export const ArticleInformation: React.FC<{}> = () => {
  const articleInfo = useSelector(getArticleInformation);
  const dispatch = useDispatch();
  const editArticleInfo = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ArticleInfoFormDialog,
        title: 'Article Information'
      })
    );
  }, [dispatch]);

  const classes = useArticleInformationStyles();
  return (
    <SectionContainer label="Article information" className={classes.sectionContainer}>
      <section className={classes.infoSection}>
        <div>
          <strong> Article DOI: </strong> {articleInfo.articleDOI}
        </div>
        <div>
          <strong> Publisher ID: </strong> {articleInfo.publisherId}
        </div>
      </section>
      <IconButton classes={{ root: classes.editButton }} onClick={editArticleInfo}>
        <EditIcon fontSize="small" />
      </IconButton>
    </SectionContainer>
  );
};
