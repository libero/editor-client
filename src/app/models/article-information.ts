import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { buildInputRules } from './plugins/input-rules';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';

import { getTextContentFromPath, makeSchemaFromConfig } from './utils';
import * as licenseTextConfig from './config/license-text.config';
import { Person } from './person';
import moment from 'moment';
import { BackmatterEntity } from './backmatter-entity';
import { JSONObject } from '../types/utility.types';

export const LICENSE_CC_BY_4 = 'CC-BY-4';
export const LICENSE_CC0 = 'CC0';

const LICENSE_CC_BY_4_TEXT = `This article is distributed under the terms of the 
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/licenses/by/4.0/">
        Creative Commons Attribution License</ext-link>, which permits unrestricted use and redistribution provided that the original author 
    and source are credited.`;

const LICENSE_CC0_TEXT = `This is an open-access article, free of all copyright, and may be freely reproduced, 
    distributed, transmitted, modified, built upon, or otherwise used by anyone for any lawful purpose.
    The work is made available under the
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/publicdomain/zero/1.0/">Creative Commons CC0 public domain dedication</ext-link>.`;

export class ArticleInformation extends BackmatterEntity {
  public static getLicenseText(licenseType: string): EditorState {
    const paragraph = document.createElement('p');
    paragraph.innerHTML = {
      [LICENSE_CC_BY_4]: LICENSE_CC_BY_4_TEXT,
      [LICENSE_CC0]: LICENSE_CC0_TEXT
    }[licenseType];

    return ArticleInformation.createLicenseEditorState(paragraph);
  }

  public static createLicenseEditorState(content?: Element): EditorState {
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    const xmlContentDocument = document.implementation.createDocument('', '', null);

    if (content) {
      xmlContentDocument.appendChild(content);
    }

    return EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor()]
    });
  }

  articleType: string;
  dtd: string;
  articleDOI: string;
  publisherId: string;
  volume: string;
  elocationId: string;
  subjects: Array<string>;
  publicationDate: string;
  licenseType: string;
  copyrightStatement: string;
  licenseText: EditorState;

  constructor(data?: JSONObject | Element, authors?: Person[]) {
    super();
    this.createEntity(data);
    if (data instanceof Element && authors) {
      this.updateCopyrightStatementFromXml(data, authors);
    }
  }

  public clone(): ArticleInformation {
    const newArticleInformation = new ArticleInformation();
    newArticleInformation.articleType = this.articleType;
    newArticleInformation.dtd = this.dtd;
    newArticleInformation.articleDOI = this.articleDOI;
    newArticleInformation.elocationId = this.elocationId;
    newArticleInformation.volume = this.volume;
    newArticleInformation.publisherId = this.publisherId;
    newArticleInformation.subjects = [...this.subjects];
    newArticleInformation.copyrightStatement = this.copyrightStatement;
    newArticleInformation.licenseType = this.licenseType;
    newArticleInformation.publicationDate = this.publicationDate;
    newArticleInformation.licenseText = this.licenseText;

    return newArticleInformation;
  }

  public updateCopyrightStatement(authors: Person[]): void {
    if (this.licenseType === LICENSE_CC_BY_4) {
      const authorNamesSection =
        authors.length === 1
          ? authors[0].lastName
          : authors.length === 2
          ? `${authors[0].lastName} and ${authors[1].lastName}`
          : authors.length > 2
          ? `${authors[0].lastName} et al`
          : '';
      const date = moment(this.publicationDate);

      this.copyrightStatement = `Â© ${date.isValid() ? date.format('YYYY') : ''}, ${authorNamesSection}`;
    } else {
      this.copyrightStatement = '';
    }
  }

  protected fromXML(xmlNode: Element): void {
    this.articleType = getTextContentFromPath(xmlNode, 'article-meta subj-group[subj-group-type="heading"]');
    const articleEl = xmlNode.querySelector('article');
    this.dtd = articleEl ? articleEl.getAttribute('dtd-version') : '';
    this.articleDOI = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="doi"]');
    this.elocationId = getTextContentFromPath(xmlNode, 'article-meta elocation-id');
    this.volume = getTextContentFromPath(xmlNode, 'article-meta volume');
    this.publisherId = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="publisher-id"]');
    this.subjects = Array.from(xmlNode.querySelectorAll('subj-group[subj-group-type="major-subject"] subject')).map(
      (el: Element) => el.textContent
    );

    this.licenseType = this.getLicenseTypeFromXml(xmlNode);
    this.publicationDate = this.getPublicationDateFromXml(xmlNode);
    this.licenseText = ArticleInformation.createLicenseEditorState(
      xmlNode.querySelector('article-meta permissions license license-p')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this.articleType = json.articleType as string;
    this.dtd = json.dtd as string;
    this.articleDOI = json.articleDOI as string;
    this.elocationId = json.elocationId as string;
    this.volume = json.volume as string;
    this.publisherId = json.publisherId as string;
    this.subjects = (json.subjects as string[]) || [];

    this.licenseType = json.licenseType as string;
    this.publicationDate = json.publicationDate as string;
    const emptyEditorState = ArticleInformation.createLicenseEditorState();
    this.licenseText = EditorState.fromJSON(
      {
        schema: emptyEditorState.schema,
        plugins: emptyEditorState.plugins
      },
      json.licenseText as JSONObject
    );
  }

  protected createBlank(): void {
    this.articleType = '';
    this.dtd = '';
    this.articleDOI = '';
    this.elocationId = '';
    this.volume = '';
    this.publisherId = '';
    this.subjects = [];

    this.licenseType = '';
    this.publicationDate = '';
    this.licenseText = ArticleInformation.createLicenseEditorState();
  }

  private getLicenseTypeFromXml(doc: Element): string {
    const licenseEl = doc.querySelector('article-meta permissions license');
    if (!licenseEl) {
      return '';
    }
    const href = licenseEl.getAttribute('xlink:href');
    if (href === 'http://creativecommons.org/licenses/by/4.0/') {
      return LICENSE_CC_BY_4;
    }
    if (href === 'http://creativecommons.org/publicdomain/zero/1.0/') {
      return LICENSE_CC0;
    }
  }

  private getPublicationDateFromXml(xmlNode: Element): string {
    let publicationDate = '';
    const pubDateNode = xmlNode.querySelector('pub-date[date-type="pub"][publication-format="electronic"]');
    if (pubDateNode) {
      publicationDate = [
        pubDateNode.querySelector('year')?.textContent,
        pubDateNode.querySelector('month')?.textContent,
        pubDateNode.querySelector('day')?.textContent
      ].join('-');
    }

    return publicationDate;
  }

  private updateCopyrightStatementFromXml(xmlNode: Element, authors: Person[]): void {
    const copyrightStatementFromXml = getTextContentFromPath(xmlNode, 'article-meta permissions copyright-statement');
    if (copyrightStatementFromXml) {
      this.copyrightStatement = copyrightStatementFromXml;
    } else {
      this.updateCopyrightStatement(authors);
    }
  }
}
