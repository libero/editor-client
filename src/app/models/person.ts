import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';
import { history } from 'prosemirror-history';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';

import * as bioConfig from 'app/models/config/author-bio.config';
import { Affiliation } from 'app/models/affiliation';
import { getTextContentFromPath, makeSchemaFromConfig } from 'app/models/utils';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Person extends BackmatterEntity {
  firstName: string;
  lastName: string;
  suffix?: string;
  isAuthenticated?: boolean;
  orcid?: string;
  email?: string;
  bio?: EditorState;
  isCorrespondingAuthor?: boolean;
  affiliations?: string[];
  hasCompetingInterest?: boolean;
  competingInterestStatement?: string;

  constructor(data?: JSONObject | Element, notesXml?: Element) {
    super();
    this.createEntity(data);
    if (data instanceof Element && notesXml instanceof Element) {
      this.setCompetingInterests(data, notesXml);
    }
  }

  public getDisplayName(): string {
    return [this.firstName, this.lastName, this.suffix].filter((_) => _).join(' ');
  }

  public getAffiliationsLabels(affiliations: Affiliation[]): string[] {
    return this.affiliations
      .map((affiliationId) => {
        const affiliation = affiliations.find(({ id }) => id === affiliationId);
        return affiliation ? affiliation.label : undefined;
      })
      .filter(Boolean)
      .sort((a, b) => Number(a) - Number(b));
  }

  clone(): Person {
    const person = new Person();
    person.firstName = this.firstName;
    person.lastName = this.lastName;
    person.suffix = this.suffix;
    person.isAuthenticated = this.isAuthenticated;
    person.email = this.email;
    person.bio = this.bio;
    person.isCorrespondingAuthor = this.isCorrespondingAuthor;
    person.affiliations = [...this.affiliations];
    person.hasCompetingInterest = this.hasCompetingInterest;
    person.competingInterestStatement = this.competingInterestStatement;

    return person;
  }

  protected fromXML(xml: Element): void {
    const orcidEl = xml.querySelector('contrib-id[contrib-id-type="orcid"]');

    this._id = xml.getAttribute('id') || this._id;
    this.firstName = getTextContentFromPath(xml, 'name > given-names');
    this.lastName = getTextContentFromPath(xml, 'name > surname');
    this.suffix = getTextContentFromPath(xml, 'name > suffix');
    this.isAuthenticated = orcidEl ? orcidEl.getAttribute('authenticated') === 'true' : false;
    this.orcid = orcidEl ? this.getOrcid(orcidEl.textContent) : '';
    this.bio = this.createBioEditorStateFromXml(xml.querySelector('bio'));
    this.email = getTextContentFromPath(xml, 'email');
    this.isCorrespondingAuthor = xml.getAttribute('corresp') === 'yes';
    this.affiliations = Array.from(xml.querySelectorAll('xref[ref-type="aff"]')).map((xRef) =>
      xRef.getAttribute('rid')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this._id = (json.id as string) || this._id;
    this.firstName = json.firstName as string;
    this.lastName = json.lastName as string;
    this.suffix = json.suffix as string;
    this.isAuthenticated = json.isAuthenticated as boolean;
    this.orcid = json.orcid as string;
    this.bio = json.bio
      ? this.createBioEditorStateFromJSON(json.bio as JSONObject)
      : this.createBioEditorStateFromXml();
    this.email = json.email as string;
    this.isCorrespondingAuthor = json.isCorrespondingAuthor as boolean;
    this.affiliations = (json.affiliations as string[]) || [];
  }

  protected createBlank(): void {
    this.firstName = '';
    this.lastName = '';
    this.suffix = '';
    this.isAuthenticated = false;
    this.orcid = '';
    this.bio = this.createBioEditorStateFromXml();
    this.email = '';
    this.isCorrespondingAuthor = false;
    this.affiliations = [];
  }

  private getOrcid(orcidUrl: string): string {
    const matches = orcidUrl.match(/(([0-9]{4}-?){4})/g);
    if (matches) {
      return matches[0];
    }
    return '';
  }

  private setCompetingInterests(dataXml: Element, notesXml: Element): void {
    const competingInterestEl = Array.from(notesXml.querySelectorAll('[fn-type="COI-statement"]')).find(
      (fnEl: Element) => {
        const id = fnEl.getAttribute('id');
        return dataXml.querySelector(`xref[ref-type="fn"][rid="${id}"]`);
      }
    );

    this.hasCompetingInterest = competingInterestEl
      ? competingInterestEl.textContent !== 'No competing interests declared'
      : false;
    this.competingInterestStatement = competingInterestEl ? competingInterestEl.textContent : '';
  }

  private createBioEditorStateFromXml(bio?: Element): EditorState {
    const schema = makeSchemaFromConfig(bioConfig.topNode, bioConfig.nodes, bioConfig.marks);
    return EditorState.create({
      doc: bio ? ProseMirrorDOMParser.fromSchema(schema).parse(bio) : undefined,
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor(), keymap(baseKeymap), history()]
    });
  }

  private createBioEditorStateFromJSON(json: JSONObject): EditorState {
    const blankState = this.createBioEditorStateFromXml();
    return EditorState.fromJSON(
      {
        schema: blankState.schema,
        plugins: blankState.plugins
      },
      json
    );
  }
}

export function createAuthorsState(authorsXml: Element[], notesXml: Element | undefined): Person[] {
  return authorsXml.map((xml) => new Person(xml, notesXml));
}
