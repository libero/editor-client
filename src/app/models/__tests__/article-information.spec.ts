import { ArticleInformation } from '../article-information';
import { Person } from '../person';

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

const AUTHORS = [
  {
    id: 'id1',
    firstName: 'First name 1',
    lastName: 'Last name 1'
  },
  {
    id: 'id2',
    firstName: 'First name 2',
    lastName: 'Last name 2'
  },
  {
    id: 'id3',
    firstName: 'First name 3',
    lastName: 'Last name 3'
  }
];
const xmlWrapper = document.createElement('div');
xmlWrapper.innerHTML = `<pub-date date-type="pub" publication-format="electronic">
    <day>30</day>
    <month>11</month>
    <year>2019</year>
  </pub-date>
  <article article-type="article-commentary" dtd-version="1.2" ></article>
  <article-meta>
    <subj-group subj-group-type="heading"> <subject>Insight</subject> </subj-group>
    <article-id pub-id-type="publisher-id">00104</article-id>
    <article-id pub-id-type="doi">10.7554/eLife.00104</article-id>
    <volume>8</volume>
    <elocation-id>e00104</elocation-id>
    <subj-group subj-group-type="major-subject"> <subject>Cell Biology</subject> </subj-group>
    <subj-group subj-group-type="major-subject"> <subject>Genetics and Genomics</subject> </subj-group>
    <permissions>
      <license xlink:href="http://creativecommons.org/licenses/by/4.0/"></license>
    </permissions>
  </article-meta>`;

describe('Article Information model helpers', () => {
  it('gets copyright statement for authors list', () => {
    let articleInfo = new ArticleInformation(xmlWrapper, [new Person(AUTHORS[0])]);
    expect(articleInfo.copyrightStatement).toBe('© 2019, Last name 1');

    articleInfo = new ArticleInformation(xmlWrapper, [new Person(AUTHORS[0]), new Person(AUTHORS[1])]);
    expect(articleInfo.copyrightStatement).toBe('© 2019, Last name 1 and Last name 2');

    articleInfo = new ArticleInformation(xmlWrapper, [
      new Person(AUTHORS[0]),
      new Person(AUTHORS[1]),
      new Person(AUTHORS[2])
    ]);
    expect(articleInfo.copyrightStatement).toBe('© 2019, Last name 1 et al');
  });

  it('creates article info state from XML with CC-BY-4', () => {
    const articleInfo = new ArticleInformation(xmlWrapper, [new Person(AUTHORS[0])]);
    expect(articleInfo).toMatchSnapshot();
  });

  it('creates article info state from XML with CC0', () => {
    const xmlWrapper = document.createElement('div');
    xmlWrapper.innerHTML = `<pub-date date-type="pub" publication-format="electronic">
        <day>30</day>
        <month>11</month>
        <year>2019</year>
      </pub-date>
      <article article-type="article-commentary" dtd-version="1.2" ></article>
      <article-meta>
        <subj-group subj-group-type="heading"> <subject>Insight</subject> </subj-group>
        <article-id pub-id-type="publisher-id">00104</article-id>
        <article-id pub-id-type="doi">10.7554/eLife.00104</article-id>
        <volume>8</volume>
        <elocation-id>e00104</elocation-id>
        <subj-group subj-group-type="major-subject"> <subject>Cell Biology</subject> </subj-group>
        <subj-group subj-group-type="major-subject"> <subject>Genetics and Genomics</subject> </subj-group>
        <permissions>
          <license xlink:href="http://creativecommons.org/publicdomain/zero/1.0/"></license>
        </permissions>
      </article-meta>`;

    const articleInfo = new ArticleInformation(xmlWrapper, [new Person(AUTHORS[0])]);
    expect(articleInfo).toMatchSnapshot();
  });
});
