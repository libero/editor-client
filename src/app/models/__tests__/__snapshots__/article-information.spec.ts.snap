// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Information model helpers creates article info state from XML with CC-BY-4 1`] = `
ArticleInformation {
  "_id": "unique_id",
  "articleDOI": "10.7554/eLife.00104",
  "articleType": " Insight ",
  "copyrightStatement": "Â© 2019, Last name 1",
  "dtd": "1.2",
  "elocationId": "e00104",
  "licenseText": Object {
    "doc": Object {
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "doc",
    },
    "selection": Object {
      "anchor": 1,
      "head": 1,
      "type": "text",
    },
  },
  "licenseType": "CC-BY-4",
  "publicationDate": "2019-11-30",
  "publisherId": "00104",
  "subjects": Array [
    "Cell Biology",
    "Genetics and Genomics",
  ],
  "volume": "8",
}
`;

exports[`Article Information model helpers creates article info state from XML with CC0 1`] = `
ArticleInformation {
  "_id": "unique_id",
  "articleDOI": "10.7554/eLife.00104",
  "articleType": " Insight ",
  "copyrightStatement": "",
  "dtd": "1.2",
  "elocationId": "e00104",
  "licenseText": Object {
    "doc": Object {
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "doc",
    },
    "selection": Object {
      "anchor": 1,
      "head": 1,
      "type": "text",
    },
  },
  "licenseType": "CC0",
  "publicationDate": "2019-11-30",
  "publisherId": "00104",
  "subjects": Array [
    "Cell Biology",
    "Genetics and Genomics",
  ],
  "volume": "8",
}
`;
